block-producer/src/block_builder/mod.rs

```diff
@@ -19,3 +45,99 @@ pub trait BlockBuilder: Send + Sync + 'static {
         batches: Vec<SharedTxBatch>,
     ) -> Result<(), BuildBlockError>;
 }
+
+#[derive(Debug)]
+pub struct DefaultBlockBuilder<S> {
+    store: Arc<S>,
+    account_root_program: AccountRootProgram,
+}
+
+impl<S> DefaultBlockBuilder<S>
+where
+    S: Store,
+{
+    pub fn new(store: Arc<S>) -> Self {
+        Self {
+            store,
+            account_root_program: AccountRootProgram::new(),
+        }
+    }
+}
+
+#[async_trait]
+impl<S> BlockBuilder for DefaultBlockBuilder<S>
+where
+    S: Store,
+{
+    async fn build_block(
+        &self,
+        batches: Vec<SharedTxBatch>,
+    ) -> Result<(), BuildBlockError> {
+        let updated_accounts: Vec<(AccountId, Digest)> =
+            batches.iter().flat_map(|batch| batch.updated_accounts()).collect();
+        let created_notes: Vec<Digest> =
+            batches.iter().flat_map(|batch| batch.created_notes()).collect();
+        let produced_nullifiers: Vec<Digest> =
+            batches.iter().flat_map(|batch| batch.produced_nullifiers()).collect();
+
+        let BlockInputs {
+            block_header: prev_block_header,
+            chain_peaks: _,
+            account_states: account_states_in_store,
+            nullifiers: _,
+        } = self
+            .store
+            .get_block_inputs(
+                updated_accounts.iter().map(|(account_id, _)| account_id),
+                produced_nullifiers.iter(),
+            )
+            .await
+            .unwrap();
+
+        let new_block_header = {
+            let prev_hash = prev_block_header.prev_hash();
+            let chain_root = Digest::default();
+            let account_root = self.account_root_program.compute_new_account_root(
+                account_states_in_store
+                    .into_iter()
+                    .map(|record| (record.account_id, record.account_hash, record.proof)),
+                updated_accounts.iter().cloned(),
+                prev_block_header.account_root(),
+            )?;
+            let nullifier_root = Digest::default();
+            let note_root = Digest::default();
+            let batch_root = Digest::default();
+            let proof_hash = Digest::default();
+            let timestamp: Felt = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .expect("today is expected to be before 1970")
+                .as_millis()
+                .into();
+
+            BlockHeader::new(
+                prev_hash,
+                prev_block_header.block_num(),
+                chain_root,
+                account_root,
+                nullifier_root,
+                note_root,
+                batch_root,
+                proof_hash,
+                prev_block_header.version(),
+                timestamp,
+            )
+        };
```

Comment:
I'd probably move this out into a separate helper function which takes `BlockWitness` as a parameter and return `BlockHeader` of the new block.

Reply:
[8d4ee51](https://github.com/0xPolygonMiden/miden-node/pull/55/commits/8d4ee51c66358db5f2812a4535ce1b3524d20472)

Moved the code block to a new function `compute_block_header()`. I currently have the function take 3 parameters instead of a new struct `BlockWitness` because I don't understand the value of using `BlockWitness`. Could you expand on why you suggested that?

Reply:
Oops, I hadn't read [this comment](https://github.com/0xPolygonMiden/miden-node/pull/55#discussion_r1385656953) where you explain `BlockWitness`. Revising...